Add the Lua Cookbook. (#8)
orocos/orocos-docs: docs/rtt/rtt-lua-cookbook.rst

Basic commands (read this!)
---------------------------

  * rttlib.stat() Print information about component instances and their state

::

    > rttlib.stat()
    Name                State               isActive  Period
    lua                 PreOperational      true      0
    Deployer            Stopped             true      0

..

  * rttlib.info() Print information about available components, types and services

::

    > rttlib.info()
    services:   marshalling scripting print LuaTLSF Lua os
    typekits:   rtt-corba-types rtt-mqueue-transport rtt-types OCLTypekit
    types:      ConnPolicy FlowStatus PropertyBag SendHandle SendStatus TaskContext array bool
                bools char double float int ints rt_string string strings uint void
    comp types: OCL::ConsoleReporting OCL::FileReporting OCL::HMIConsoleOutput OCL::HelloWorld
                OCL::LuaComponent OCL::LuaTLSFComponent OCL::TcpReporting
    ...

..

Where's my TaskContext?
-----------------------

Here:

::

    > tc = rtt.getTC()

..

Above code calls the getTC() function, which returns the current TC and stores
it in a variable 'tc'. For showing the interface just write =tc. In the repl

> the equal sign is a shortcut for 'return', which in turn causes the variable to

Reviewer: ```suggestion
Above code calls the ``getTC()`` function, which returns the current TC and stores
it in a variable ``tc``. For showing the interface just write ``=tc``. In the repl
the equal sign is a shortcut for ``return``, which in turn causes the variable to

be printed. (BTW: This works for displaying any variable)

::

    > =tc
    TaskContext: lua
    state: PreOperational
    isActive: true
    getPeriod: 0
    peers: Deployer
    ports:
    properties:
      lua_string (string) =  // string of lua code to be executed during configureHook
      lua_file (string) =  // file with lua program to be executed during configuration
    operations:
      bool exec_file(string const& filename) // load (and run) the given lua script
      bool exec_str(string const& lua-string) // evaluate the given string in the lua environment

..

Since (rttlua beta5) the above does not print the standard TaskContext operations

> anymore. To print these, use tc:show().

Reviewer: ```suggestion
anymore. To print these, use ``tc:show()``.
```

Getting persistent history with rlwrap
--------------------------------------

rttlua does not offer persistent history like in the taskbrowser. If you want it,
 you can use rlwrap and to wrap rttlua as follows:

::

    alias rttlua='rlwrap -a -r -H ~/.rttlua-history rttlua-gnulinux'

..

> If you run 'rttlua' it should have persistent history.

Reviewer: ```suggestion
If you run ``rttlua`` it should have persistent history.
```

Dataflow
========

The following shows the basic API, see section Automatically creating and cleaning
up component interfaces for a more convenient way add/remove ports/properties.

Creating Ports
--------------

::

    > pin = rtt.InputPort("string")
    > pout = rtt.OutputPort("string")
    > =pin
     [in, string, unconn, local] //
    > =pout
     [out, string, unconn, local] //

..

Both In- and OutputPorts optionally take a second string argument (name) and third
argument (description).

Connecting Ports
----------------

Directly
^^^^^^^^

> For this the ports don't have to be added to the TaskContext:

Reviewer: ```suggestion
For this the ports don't have to be added to the ``TaskContext``:
```

::

    > =pin:connect(pout)
    true
    > return pin
     [in, string, conn, local] //
    > return pout
     [out, string, conn, local] //
    >

..

Using the Deployer
^^^^^^^^^^^^^^^^^^

> The rttlua-* REPL automatically creates a deployment component that is a peer of the lua taskcontext:

Reviewer: ```suggestion
The ``rttlua-*`` REPL automatically creates a deployment component that is a peer of the lua taskcontext:
```

                    OCL::logging::LoggingService, OCL::logging::OstreamAppender, TaskContext

..

Creating RTT types
------------------

::

    > cp = rtt.Variable("ConnPolicy")
    > =cp
    {data_size=0,type="DATA",name_id="",init=false,pull=false,transport="default",lock_policy="LOCK_FREE",size=0}
    > cp.data_size = 4711
    > print(cp.data_size)
    4711

..

Accessing global RTT constants
------------------------------

*Printing the available constants:*

::

    > =rtt.globals
    {SendNotReady=SendNotReady,LOCK_FREE=2,NewData=NewData,SendFailure=SendFailure,\
    SendSuccess=SendSuccess,NoData=NoData,UNSYNC=0,LOCKED=1,OldData=OldData,BUFFER=1,DATA=0}
    >

..

*Accessing constants - just index!*

::

  > =rtt.globals.LOCK_FREE
  2

..

Convenient initalization of multi-field types
---------------------------------------------

It is cumbersome to initalize complex types with many subfields:

::

    > tc = rtt.getTC()
    > depl = tc:getPeer("Deployer")
    > depl:import("kdl_typekit")
    > t=rtt.Variable("KDL.Frame")
    > =t
    {M={Z_y=0,Y_y=1,X_y=0,Y_z=0,Z_z=1,Y_x=0,Z_x=0,X_x=1,X_z=0},p={Y=0,X=0,Z=0}}
    > t.M.X_x=3
    > t.M.Y_x=2
    > t.M.Z_x=2.3
    ...

..

> To avoid this, use the fromtab() method:

Reviewer: ```suggestion
To avoid this, use the ``fromtab()`` method:
```

..

(Note: the second and third argument (name and description) are optional and
can also be set when adding the property to a TaskContext)

Adding to TaskContext Interface
-------------------------------

::

  > tc=rtt.getTC()
  > tc:addProperty(p1)
  > =tc -- check it is there...

..

Getting a Properties from a TaskContext
---------------------------------------

::

    > tc=rtt.getTC()
    > pgain = tc:getProperty("pgain")
    > =pgain -- will print it

..

Properties of basic types: setting the value
--------------------------------------------

::

    > p1:set(3.14)
    > =p1  -- a property can be printed!
    p-gain (double) = 3.14 // Proportional controller gain

..

In particular, the following will not work:

::

    > p1=3.14

..

Lua works with references! This will assign the variable p1 a numeric value of 3.14
and the reference to the property is lost.

Properties of basic types: getting the value
--------------------------------------------

::

    > print("the value of " .. p1:info().name .. " is: " .. p1:get())
    the value of p-gain is: 3.14

..

Properties of complex types: accessing
--------------------------------------

> Assume a property of type KDL::Frame. Similarily to Variables the subfields

Reviewer: ```suggestion
Assume a property of type ``KDL::Frame``. Similarily to Variables the subfields
```

::

    > tc:removeProperty("p-gain")
    > =tc         -- p-gain is gone now
    > p1:delete() -- delete property and free memory
    > =p1         -- p1 is 'dead' now.
    userdata: 0x186f8c8

..

Operations
==========

Synchronous calling of operations from Lua:

Calling Operations
------------------

The short and convenient way
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

    > d = tc:getPeer("Deployer")
    > =d:getPeriod()
    0

..

The significantly faster and real-time safe way (because locally cached)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

::

    > d = tc:getPeer("Deployer")
    > op = d:getOperation("getPeriod")
    > =op -- can be printed!
    double getPeriod() // Get the configured execution period. -1.0: no thread ...
    > =op() -- call it
    0

..

Sending Operations
------------------

"Sending" Operations permits to asynchronously request an operation to be executed and collect the results at a later point in time.

::

    > d = tc:getPeer("Deployer")
    > op = d:getOperation("getPeriod")
    > handle=op:send() -- calling it
    > =handle:collect()
    SendSuccess    0

..

.. note::

>     - ``collect()`` returns multiple arguments: first a SendStatus string ('SendSuccess', 'SendFailure') followed by zero to many output arguments of the operation.
>     - ``collect`` blocks until the operation was executed, collectIfDone() will immediately return (but possibly with 'SendNotReady')

Reviewer: ```suggestion
    - ``collect()`` returns multiple arguments: first a SendStatus string (``SendSuccess``, ``SendFailure``) followed by zero to many output arguments of the operation.
        - ``collect`` blocks until the operation was executed, ``collectIfDone()`` will immediately return (but possibly with ``SendNotReady``)
```

    tc = rtt.getTC()
    depl = tc:getPeer("Deployer")

    -- import components, requires correctly setup RTT_COMPONENT_PATH
    depl:import("ocl")
    -- depl:import("componentX")
    -- import components, requires correctly setup ROS_PACKAGE_PATH (>=Orocos 2.7)
    depl:import("rtt_ros")
    rtt.provides("ros"):import("my_ros_pkg")

    -- create component 'hello'
    depl:loadComponent("hello", "OCL::HelloWorld")

    -- get reference to new peer
    hello = depl:getPeer("hello")

    -- create buffered connection of size 64
    cp = rtt.Variable('ConnPolicy')
    cp.type=1   -- type buffered
    cp.size=64  -- buffer size
    depl:connect("hello.the_results", "hello.the_buffer_port", cp)
    rtt.logl('Info', "Deployment complete!")

..

run it:

::

    $ rttlua-gnulinux -i deploy-app.lua

..

or using orocos_toolchain_ros

::

    $ rosrun ocl rttlua-gnulinux -i deploy-app.lua

..

.. note::

    The -i option makes rttlua enter interactive mode (the REPL) after executing the script. Without it would exit after finishing executing the script, which in this case is probably not what you want.

..

How to write a RTT-Lua component
--------------------------------

> A Lua component is created by loading a Lua-script implementing zero or more TaskContext hooks in a OCL::LuaComponent. The following RTT hooks are currently supported:

Reviewer: ```suggestion
A Lua component is created by loading a Lua-script implementing zero or more ``TaskContext`` hooks in a ``OCL::LuaComponent``. The following RTT hooks are currently supported:
```

    -- load Lua service into the HelloWorld Component
    d:loadService("hello", "Lua")

    -- Execute the following Lua script (defined a multiline string) in
    -- the service. This dummy examples simply modifies the Property.  For
    -- large programs it might be better tostore the program in a separate
    -- file and use the exec_file operation instead.
    proggie = [[
        require("rttlib")
        tc=rtt.getTC() -- this is the Hello Component
        prop = tc:getProperty("the_property")
        prop:set("hullo from the lua service!")
    ]]

    prop = hello:getProperty("the_property") -- get hello.the_property
    print("the_property before service call:", prop)
    hello:provides("Lua"):exec_str(proggie) -- execute program in the service
    print("the_property after service call: ", prop)

..

Executing a LuaService function at the frequency of the host component
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
More useful than just running once is to be able to execute a function synchronously with the updateHook of the host component. This can be achieved by registering a ExecutionEngine hook (much easier than it sounds!).

> The following Lua service code implements a simple monitor that tracks the currently active (TaskContext) state of the component in whose context it is running. When the state changes the new state is written to a port "tc_state", which is added to the context TC.

Reviewer: ```suggestion
The following Lua service code implements a simple monitor that tracks the currently active (``TaskContext``) state of the component in whose context it is running. When the state changes the new state is written to a port ``tc_state``, which is added to the context TC.
```

    hello = d:getPeer("hello")

    -- load Lua service into the HelloWorld Component
    d:loadService("hello", "Lua")

    mon_state = [[
        -- service-eehook.lua
        require("rttlib")
        tc=rtt.getTC() -- this is the Hello Component
        last_state = "not-running"
        out = rtt.OutputPort("string")
        tc:addPort(out, "tc_state", "currently active state of TaskContext")

        function check_state()
            local cur_state = tc:getState()
            if cur_state ~= last_state then
                out:write(cur_state)
                last_state = cur_state
            end
            return true -- returning false will disable EEHook
        end

        -- register check_state function to be called periodically and
        -- enable it. Important: variables like eehook below or the
        -- function check_state which shall not be garbage-collected
        -- after the first run must be declared global (by not declaring
        -- them local with the local keyword)
        eehook=rtt.EEHook('check_state')
        eehook:enable()
    ]]

    -- execute the mon_state program
    hello:provides("Lua"):exec_str(mon_state)

..

.. note::

>     the -i option causes rttlua to go to interactive mode after executing the script (and not exiting afterwards).

Reviewer: ```suggestion
    the ``-i`` option causes ``rttlua`` to go to interactive mode after executing the script (and not exiting afterwards).
```

..

::

    $ rttlua-gnulinux -i service-eehook.lua
    > rttlib.portstats(hello)
    the_results (string)  =
    the_buffer_port (string)  = NoData
    tc_state (string)  = Running
    > hello:error()
    > rttlib.portstats(hello)
    the_results (string)  =
    the_buffer_port (string)  = NoData
    tc_state (string)  = RunTimeError
    >

..

How to perform runtime system validation?
-----------------------------------------

It is often useful to validate a deployed system at runtime, however you want to avoid cluttering individual components with non-functional validation code. Here's what to do (Please also see this post on orocos-users, which inspired the following)

**Use-case**: check for unconnected input ports

1. Write a function to validate a **single component**

> The following function accepts a TaskContext as an argument and checks wether it has unconnected input ports. If yes it prints an error.

Reviewer: ```suggestion
The following function accepts a ``TaskContext`` as an argument and checks whether it has unconnected input ports. If yes it prints an error.
```

.. code-block:: lua

    function check_inport_conn(tc)
    local portnames = tc:getPortNames()
    local ret = true
    for _,pn in ipairs(portnames) do
        local p = tc:getPort(pn)
        local info = p:info()
        if info.porttype == 'in' and info.connected == false then
            rtt.logl('Error', "InputPort " .. tc:getName() .. "." .. info.name .. " is unconnected!")
            ret = false
        end
    end
    return ret
    end

..

2. After deployment, **execute the validation function on all components**:

This can be done using the ``mappeers`` function.

.. code-block:: bash

    rttlib.mappeers(check_inport_conn, depl)

..

> The ``mappeers`` function is a special variant of map which calls the function given as a first argument on all peers reachable from a TaskContext (given as a second argument). We pass the Deployer here, which typically knows all components.

Reviewer: ```suggestion
The ``mappeers`` function is a special variant of map which calls the function given as a first argument on all peers reachable from a ``TaskContext`` (given as a second argument). We pass the Deployer here, which typically knows all components.
```

Where to run a statemachine: Component vs. Service?
---------------------------------------------------

Answer:

Typically a Component will be preferred when

    - the statemachine has to coordinate/interact with/supervise multiple components
    - it shall run purely event-driven or at a different frequency than the computational components

A Service is preferred when

    - the Statemachine coordinates/monitors only one component
    - the Statemachine runs synchronous (same frequency) with the host component

There will, undoubtly, be exceptions!

How to run an rFSM in a Component
---------------------------------

Summary: Create a OCL::LuaComponent. In ``configureHook`` load and initalize the fsm, in ``updateHook`` call  ``rfsm.run(fsm)``

(see the rFSM docs for general information)

It is a best-practice to split the initalization (setting up required functions, peers or ports used by the fsm) and the fsm model itself into two files. This way the fsm model is kept as platform independent and hence reusable as possible.

> The following initalization file is executed in the newly create LuaComponent for preparing the environment for the statemachine, that is loaded and initalized in configureHook.

Reviewer: ```suggestion
The following initalization file is executed in the newly create LuaComponent for preparing the environment for the statemachine, that is loaded and initalized in ``configureHook``.
```

        rfsm.post_step_hook_add(fsm, rfsm_rtt.gen_write_fqn(fqn_out))
        return true
    end

    function updateHook() rfsm.run(fsm) end

    function cleanupHook()
        -- cleanup the created ports.
        rttlib.tc_cleanup()
    end

..

A dummy statemachine stored in the **fsm.lua** file:

.. code-block:: lua

    return rfsm.state {
        ping = rfsm.state {
            entry=function() print("in ping entry") end,
        },

        pong = rfsm.state {
            entry=function() print("in pong entry") end,
        },

        rfsm.trans {src="initial", tgt="ping" },
        rfsm.trans {src="ping", tgt="pong", events={"e_pong"}},
        rfsm.trans {src="pong", tgt="ping", events={"e_ping"}},

..

**Option A: Running the rFSM example with a Lua deployment script**

**deploy.lua**

.. code-block:: lua

    -- alternate lua deploy script
    require "rttlib"

    tc=rtt.getTC()
    d=tc:getPeer("Deployer")

    d:import("ocl")
    d:loadComponent("Supervisor", "OCL::LuaComponent")
    sup = d:getPeer("Supervisor")

    sup:exec_file("launch_fsm.lua")
    sup:configure()
    cmd = rttlib.port_clone_conn(sup:getPort("events"))

..

> Run it. cmd is an inverse (output) port which is connected to the incoming (from POV of the fsm) 'events' port of the fsm, so by writing to it we can send events:

Reviewer: ```suggestion
Run it. ``cmd`` is an inverse (output) port which is connected to the incoming (from POV of the fsm) ``events`` port of the fsm, so by writing to it we can send events:
```

.. code-block:: bash

    $ rosrun ocl rttlua-gnulinux -i deploy.lua
    OROCOS RTTLua 1.0-beta3 / Lua 5.1.4 (gnulinux)
    INFO: created undeclared connector root.initial
    > sup:start()
    > in ping entry

    > cmd:write("e_pong")
    > in pong entry

    > cmd:write("e_ping")
    > in ping entry

    > cmd:write("e_pong")
    > in pong entry

..

**Option B: Running the rFSM example with an Orocos deployment script**

**deploy.ops**

::

    import("ocl")
    loadComponent("Supervisor", "OCL::LuaComponent")
    Supervisor.exec_file("launch_fsm.lua")
    Supervisor.configure

..

> After starting the supervisor we 'leave' it, so we can write to the 'events' ports:

Reviewer: ```suggestion
After starting the supervisor we ``leave`` it, so we can write to the ``events`` ports:
```

.. code-block:: bash

    $ rosrun ocl deployer-gnulinux -s deploy.ops
    INFO: created undeclared connector root.initial
    Switched to : Deployer

        This console reader allows you to browse and manipulate TaskContexts.
        You can type in an operation, expression, create or change variables.
        (type 'help' for instructions and 'ls' for context info)

            TAB completion and HISTORY is available ('bash' like)

    Deployer [S]> cd Supervisor

    TaskBrowser connects to all data ports of Supervisor
    Switched to : Supervisor
    Supervisor [S]> start
    = true

    Supervisor [R]> in ping entry

    Supervisor [R]> leave
    Watching Supervisor [R]> events.write ("e_pong")
    = (void)

    Watching Supervisor [R]> in pong entry

    Watching Supervisor [R]> events.write ("e_ping")
    = (void)

    Watching Supervisor [R]> in ping entry

    Watching Supervisor [R]>

..

Running rFSM in a Service
-------------------------

This is basically the same as executing a function periodally in a service (see the Service example above). There is a convenience function service_launch_rfsm in rfsm_rtt.lua to make this easier.

The steps are:

>     - create LuaService in Component in question
>     - prepare Lua environment, i.e. call exec_string or exec_file to add functions.

Reviewer: ```suggestion
    - create ``LuaService`` in Component in question
    - prepare Lua environment, i.e. call ``exec_string`` or ``exec_file`` to add functions.
```

    - launch the fsm with the following call in your deployment script:

.. code-block:: lua

    require "rfsm_rtt"

    -- get reference to exec_str operation
    fsmfile = "fsm.lua"
    execstr_op = comp:provides("Lua"):getOperation("exec_str")
    rfsm_rtt.service_launch_rfsm(fsmfile, execstr_op, true)

..

> The last line means the following: launch fsm in <fsmfile> in service identified by execstr_op, true: create an execution engine hook so that the rfsm.step is called at the component frequency. (See the generated rfsm_rtt API docs).

Reviewer: ```suggestion
The last line means the following: launch fsm in ``<fsmfile>`` in service identified by ``execstr_op``, true: create an execution engine hook so that the ``rfsm.step`` is called at the component frequency. (See the generated ``rfsm_rtt`` API docs).
```

..

Assume the Deployer has two peers "robot" and "controller", they can be accessed as follows:

::

    print(peertab.robot)
    -- or
    peertab.controller:configure()

..

Miscellaneous
=============

Connecting RTT Ports to ROS topics
----------------------------------

::

    > cp=rtt.Variable("ConnPolicy")
    > cp.transport=3 -- 3 is ROS
    > cp.name_id="/l_cart_twist/command" -- topic name
    > depl:stream("CompX.portY", cp)

..

or with sweet one-liner (thx to Ruben!):

::

    > depl:stream("CompX.portY", rtt.provides("ros"):topic("/l_cart_twist/command"))

..

Finding the path to a ROS package
---------------------------------

This is sometimes usefull for loading scripts etc. that are located in different packages.

The rttros.lua collects some basic but useful stuff for interacting with ROS. This one is "borrowed" from the excellent roslua:

::

    > require "rttros"
    > =rttros.find_rospack("geometry_msgs")
    /home/mk/src/ros/unstable/common_msgs/geometry_msgs
    >

..

How are types converted between RTT and Lua?
--------------------------------------------

+--------+--------+
| RTT    | Lua    |

> +========+========+
> | bool 	 | boolean|
> +--------+--------+

Reviewer: ```suggestion
+========+========+
| bool   | boolean|
+--------+--------+
```

**Quick example**: ``ConnPolicy`` type

(This is just an example. It has been done for this type already).

The out-of-box printing of a ConnPolicy will look as follows:

::

    ./rttlua-gnulinux
    Orocos RTTLua 1.0-beta3 (gnulinux)
    > return rtt.Variable("ConnPolicy")
    {data_size=0,type=0,name_id="",init=false,pull=false,transport=0,lock_policy=2,size=0}

..

This not too bad, but we would like to display the string representation of the C++ enums ``type`` and ``lock_policy``. So we must write a function that returns a table...

.. code-block:: lua

    function ConnPolicy2tab(cp)
        if cp.type == 0 then cp.type = "DATA"
        elseif cp.type == 1 then cp.type = "BUFFER"
        else cp.type = tostring(cp.type) .. " (invalid!)" end

        if cp.lock_policy == 0 then cp.lock_policy = "UNSYNC"
        elseif cp.lock_policy == 1 then cp.lock_policy = "LOCKED"
        elseif cp.lock_policy == 2 then cp.lock_policy = "LOCK_FREE"
        else cp.lock_policy = tostring(cp.lock_policy) .. " (invalid!)" end
        return cp
    end

..

> and add it to the `rttlib.var_pp`` table of Variable formatters as follows:

Reviewer: ```suggestion
and add it to the ``rttlib.var_pp`` table of Variable formatters as follows:
```

::

    rttlib.var_pp.ConnPolicy = ConnPolicy2tab

..

now printing a ``ConnPolicy`` again calls our function and prints the desired fields:

::

    > return rtt.Variable("ConnPolicy")
    {data_size=0,type="DATA",name_id="",init=false,pull=false,transport=0,lock_policy="LOCK_FREE",size=0}
    >

..

How to use classical OCL Deployers ? (like with Corba, or with a Taskbrowser)
-----------------------------------------------------------------------------

> If you are used to manage your application with the classic OCL Taskbrowser or if you want your application to be connected via Corba, you may only use lua for deployment, and continue to use your former deployer. To do so, you have to load the lua service into your favorite deployer (deployer, cdeployer, deployer-corba, ...) and then call your deployment script.

Reviewer: ```suggestion
If you are used to manage your application with the classic ``OCL::Taskbrowser`` or if you want your application to be connected via Corba, you may only use lua for deployment, and continue to use your former deployer. To do so, you have to load the lua service into your favorite deployer (``deployer``, ``cdeployer``, ``deployer-corba``, ...) and then call your deployment script.
```

Exemple : launch your prefered deployer :

::

    cdeployer -s loadLua.ops

..

with loadLua.ops :

::

    //load the lua service
    loadService ("Deployer","Lua")

    //execute your deployment file
    Lua.exec_file("yourLuaDeploymentFile.lua")

..

and with yourLuaDeploymentFile.lua containing the kind of stuff described in this Cookbook. Like the one in paragraph "How to write a deployment script"

How to generate graphical representations of rFSM models
--------------------------------------------------------

> The rfsm-viz command allows you to generate easy-to-read pictures representing the structure of your FSM model. This tool uses the rfsm2uml and fsm2dbg modules and requires the libgv-lua package. Practically:

Reviewer: ```suggestion
The ```rfsm-viz``` command allows you to generate easy-to-read pictures representing the structure of your FSM model. This tool uses the ``rfsm2uml`` and ``fsm2dbg`` modules and requires the ``libgv-lua`` package. Practically:
```
